File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shaders/uniforms/TextureUniform.java
package org.saar.lwjgl.opengl.shaders.uniforms;

import org.lwjgl.opengl.GL20;
import org.saar.lwjgl.opengl.textures.ReadOnlyTexture;
import org.saar.lwjgl.opengl.textures.Texture;

public abstract class TextureUniform extends UniformBase implements Uniform {

    @Override
    public void doInitialize() {
        GL20.glUniform1i(getLocation(), getUnit());
    }

    @Override
    public final void doLoad() {
        Texture.bind(getUniformValue(), getUnit());
    }

    public abstract ReadOnlyTexture getUniformValue();

    public abstract int getUnit();
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/shader/uniforms/TextureUniform.kt
package org.saar.lwjgl.opengl.shader.uniforms

import org.lwjgl.opengl.GL20
import org.saar.lwjgl.opengl.texture.ReadOnlyTexture

abstract class TextureUniform : Uniform {

    final override fun load(location: Int) {
        GL20.glUniform1i(location, this.unit)
        this.value.bind(this.unit)
    }

    abstract val value: ReadOnlyTexture
    abstract val unit: Int
}